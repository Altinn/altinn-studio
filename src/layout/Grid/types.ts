import type { GridCellLabelFrom, GridCellText, GridComponentRef, GridRow } from 'src/layout/common.generated';

export interface GridCellNode extends Omit<GridComponentRef, 'component'> {
  nodeId: string;
}

export type GridCellInternal = GridCellNode | null | GridCellText | GridCellLabelFrom;

export interface GridRowInternal extends Omit<GridRow, 'cells'> {
  cells: GridCellInternal[];
}

export type GridRowsInternal = GridRowInternal[];
