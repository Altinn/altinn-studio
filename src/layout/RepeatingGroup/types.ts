import type { ExprResolved } from 'src/features/expressions/types';
import type { GridRows } from 'src/layout/common.generated';
import type { RepeatingGroupDef } from 'src/layout/RepeatingGroup/config.def.generated';

type Comp = ReturnType<RepeatingGroupDef['evalDefaultExpressions']>;
type RepGroupTrb = Exclude<Comp['textResourceBindings'], undefined>;
type RepGroupEdit = Exclude<Comp['edit'], undefined>;

// These types define the properties in a repeating group config that will have their expressions resolved
// per row instead of for the entire repeating group component at once.
type PerRowTrb =
  | 'save_and_next_button'
  | 'save_button'
  | 'edit_button_close'
  | 'edit_button_open'
  | 'multipage_next_button'
  | 'multipage_back_button';
type PerRowEdit = 'deleteButton' | 'saveButton' | 'editButton' | 'alertOnDelete' | 'saveAndNextButton';
export type GroupExpressions = ExprResolved<{
  textResourceBindings?: Pick<RepGroupTrb, PerRowTrb>;
  edit?: Pick<RepGroupEdit, PerRowEdit>;
}>;

// This then, by its definition, is the opposite of the above types. It's the properties that are resolved for the
// entire repeating group component at once.
type RepGroupBase = ExprResolved<
  Omit<Comp, 'hiddenRow' | 'textResourceBindings' | 'edit' | 'rowsAfter' | 'rowsBefore'>
>;
export type RepGroupInternal = RepGroupBase & {
  textResourceBindings?: Omit<RepGroupTrb, PerRowTrb>;
  edit?: Omit<RepGroupEdit, PerRowEdit>;
  rowsBefore?: GridRows;
  rowsAfter?: GridRows;
};
