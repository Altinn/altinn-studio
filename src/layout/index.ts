import { createContext } from 'react';
import type React from 'react';

import { Address } from 'src/layout/Address/index';
import { AttachmentList } from 'src/layout/AttachmentList/index';
import { Button } from 'src/layout/Button/index';
import { Checkboxes } from 'src/layout/Checkboxes/index';
import { Custom } from 'src/layout/Custom/index';
import { Datepicker } from 'src/layout/Datepicker/index';
import { Dropdown } from 'src/layout/Dropdown/index';
import { FileUpload } from 'src/layout/FileUpload/index';
import { FileUploadWithTag } from 'src/layout/FileUploadWithTag/index';
import { Header } from 'src/layout/Header/index';
import { Image } from 'src/layout/Image/index';
import { Input } from 'src/layout/Input/index';
import { InstanceInformation } from 'src/layout/InstanceInformation/index';
import { InstantiationButton } from 'src/layout/InstantiationButton/index';
import { Likert } from 'src/layout/Likert/index';
import { List } from 'src/layout/List/index';
import { Map } from 'src/layout/Map/index';
import { MultipleSelect } from 'src/layout/MultipleSelect/index';
import { NavigationBar } from 'src/layout/NavigationBar/index';
import { NavigationButtons } from 'src/layout/NavigationButtons/index';
import { Panel } from 'src/layout/Panel/index';
import { Paragraph } from 'src/layout/Paragraph/index';
import { PrintButton } from 'src/layout/PrintButton/index';
import { RadioButtons } from 'src/layout/RadioButtons/index';
import { TextArea } from 'src/layout/TextArea/index';
import type { ExprResolved } from 'src/features/expressions/types';
import type { IGenericComponentProps } from 'src/layout/GenericComponent';
import type { ComponentExceptGroup, ComponentExceptGroupAndSummary, IGrid, ILayoutComponent } from 'src/layout/layout';
import type { LayoutComponent } from 'src/layout/LayoutComponent';
import type { IComponentValidations } from 'src/types';
import type { ILanguage } from 'src/types/shared';
import type { IComponentFormData } from 'src/utils/formComponentUtils';

export const components: {
  [Type in ComponentExceptGroupAndSummary]: LayoutComponent<Type>;
} = {
  AddressComponent: new Address(),
  AttachmentList: new AttachmentList(),
  Button: new Button(),
  Checkboxes: new Checkboxes(),
  Custom: new Custom(),
  Datepicker: new Datepicker(),
  Dropdown: new Dropdown(),
  FileUpload: new FileUpload(),
  FileUploadWithTag: new FileUploadWithTag(),
  Header: new Header(),
  Image: new Image(),
  Input: new Input(),
  InstanceInformation: new InstanceInformation(),
  InstantiationButton: new InstantiationButton(),
  Likert: new Likert(),
  Map: new Map(),
  MultipleSelect: new MultipleSelect(),
  NavigationBar: new NavigationBar(),
  NavigationButtons: new NavigationButtons(),
  Panel: new Panel(),
  Paragraph: new Paragraph(),
  PrintButton: new PrintButton(),
  RadioButtons: new RadioButtons(),
  TextArea: new TextArea(),
  List: new List(),
};

export interface IComponentProps extends IGenericComponentProps {
  handleDataChange: (
    value: string | undefined,
    options?: {
      key?: string; // Defaults to simpleBinding
      validate?: boolean; // Defaults to true
    },
  ) => void;
  getTextResource: (key: string) => React.ReactNode;
  getTextResourceAsString: (key: string) => string;
  language: ILanguage;
  shouldFocus: boolean;
  text: React.ReactNode | string;
  label: () => JSX.Element;
  legend: () => JSX.Element;
  formData: IComponentFormData;
  isValid?: boolean;
  componentValidations?: IComponentValidations;
}

export type PropsFromGenericComponent<T extends ComponentExceptGroup> = IComponentProps &
  ExprResolved<Omit<ILayoutComponent<T>, 'type'>>;

export interface IFormComponentContext {
  grid?: IGrid;
  id?: string;
  baseComponentId?: string;
}

export const FormComponentContext = createContext<IFormComponentContext>({
  grid: undefined,
  id: undefined,
  baseComponentId: undefined,
});

/**
 * This enum is used to distinguish purly presentational components
 * from interactive form components that can have formData etc.
 */
export enum ComponentType {
  Presentation = 'presentation',
  Form = 'form',
  Button = 'button',
}
