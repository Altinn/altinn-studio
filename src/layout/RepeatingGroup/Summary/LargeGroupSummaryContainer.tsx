import React from 'react';
import type { JSX } from 'react';

import { Fieldset, Heading } from '@digdir/designsystemet-react';
import cn from 'classnames';

import { Lang } from 'src/features/language/Lang';
import classes from 'src/layout/RepeatingGroup/Summary/LargeGroupSummaryContainer.module.css';
import { pageBreakStyles } from 'src/utils/formComponentUtils';
import { BaseLayoutNode } from 'src/utils/layout/LayoutNode';
import { Hidden } from 'src/utils/layout/NodesContext';
import { useNodeDirectChildren, useNodeItem } from 'src/utils/layout/useNodeItem';
import { useNodeTraversal } from 'src/utils/layout/useNodeTraversal';
import type { HeadingLevel } from 'src/layout/common.generated';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';
import type { TraversalRestriction } from 'src/utils/layout/useNodeTraversal';

export interface IDisplayRepAsLargeGroup {
  groupNode: LayoutNode<'RepeatingGroup'>;
  id?: string;
  restriction?: TraversalRestriction;
  renderLayoutNode: (node: LayoutNode) => JSX.Element | null;
}

const headingSizes: { [k in HeadingLevel]: Parameters<typeof Heading>[0]['size'] } = {
  [2]: 'medium',
  [3]: 'small',
  [4]: 'xsmall',
  [5]: 'xsmall',
  [6]: 'xsmall',
};

export function LargeGroupSummaryContainer({ groupNode, id, restriction, renderLayoutNode }: IDisplayRepAsLargeGroup) {
  const item = useNodeItem(groupNode);
  const isHidden = Hidden.useIsHidden(groupNode);
  const depth = useNodeTraversal((t) => t.parents().length, groupNode);
  const children = useNodeDirectChildren(groupNode, restriction);
  if (isHidden) {
    return null;
  }
  const { title, summaryTitle } = item.textResourceBindings || {};

  const isNested = groupNode.parent instanceof BaseLayoutNode;
  const headingLevel = Math.min(Math.max(depth + 1, 2), 6) as HeadingLevel;
  const headingSize = headingSizes[headingLevel];
  const legend = summaryTitle ?? title;

  return (
    <Fieldset
      legend={
        legend && (
          <Heading
            level={headingLevel}
            size={headingSize}
          >
            <Lang
              id={legend}
              node={groupNode}
            />
          </Heading>
        )
      }
      className={cn(pageBreakStyles(item.pageBreak), classes.summary, {
        [classes.largeGroupContainer]: !isNested,
      })}
    >
      <div
        id={id || item.id}
        className={classes.largeGroupContainer}
      >
        {children.map((n) => renderLayoutNode(n))}
      </div>
    </Fieldset>
  );
}
