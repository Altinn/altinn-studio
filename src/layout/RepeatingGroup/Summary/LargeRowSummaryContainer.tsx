import React from 'react';
import type { JSX } from 'react';

import { Fieldset, Heading } from '@digdir/designsystemet-react';
import cn from 'classnames';

import { useLayoutLookups } from 'src/features/form/layout/LayoutsContext';
import { Lang } from 'src/features/language/Lang';
import classes from 'src/layout/RepeatingGroup/Summary/LargeGroupSummaryContainer.module.css';
import { RepGroupHooks } from 'src/layout/RepeatingGroup/utils';
import { pageBreakStyles } from 'src/utils/formComponentUtils';
import { useComponentIdMutator, useIndexedId } from 'src/utils/layout/DataModelLocation';
import { Hidden, NodesInternal } from 'src/utils/layout/NodesContext';
import { useItemWhenType } from 'src/utils/layout/useNodeItem';
import type { HeadingLevel } from 'src/layout/common.generated';

export interface IDisplayRepAsLargeGroup {
  baseComponentId: string;
  id?: string;
  renderLayoutComponent: (baseId: string) => JSX.Element | null;
  inExcludedChildren: (indexedId: string, baseId: string) => boolean;
}

const headingSizes: { [k in HeadingLevel]: Parameters<typeof Heading>[0]['data-size'] } = {
  [2]: 'md',
  [3]: 'sm',
  [4]: 'xs',
  [5]: 'xs',
  [6]: 'xs',
};

export function LargeRowSummaryContainer({
  baseComponentId,
  id,
  renderLayoutComponent,
  inExcludedChildren,
}: IDisplayRepAsLargeGroup) {
  const item = useItemWhenType(baseComponentId, 'RepeatingGroup');
  const isHidden = Hidden.useIsHiddenSelector();
  const indexedId = useIndexedId(baseComponentId, true);
  const depth = NodesInternal.useSelector((state) => state.nodeData?.[indexedId]?.depth);
  const layoutLookups = useLayoutLookups();
  const children = RepGroupHooks.useChildIds(baseComponentId);
  const idMutator = useComponentIdMutator();

  if (typeof depth !== 'number') {
    return null;
  }

  const { title, summaryTitle } = item.textResourceBindings || {};
  const parent = layoutLookups.componentToParent[baseComponentId];
  const isNested = parent?.type === 'node';
  const headingLevel = Math.min(Math.max(depth + 1, 2), 6) as HeadingLevel;
  const headingSize = headingSizes[headingLevel];
  const legend = summaryTitle ?? title;

  return (
    <Fieldset
      className={cn(pageBreakStyles(item.pageBreak), classes.summary, {
        [classes.largeGroupContainer]: !isNested,
      })}
    >
      <Fieldset.Legend>
        <Heading
          level={headingLevel}
          data-size={headingSize}
        >
          <Lang id={legend} />
        </Heading>
      </Fieldset.Legend>
      <div
        id={id || item.id}
        className={classes.largeGroupContainer}
      >
        {children.map((id) => {
          if (inExcludedChildren(idMutator(id), id) || isHidden(idMutator(id), 'node')) {
            return null;
          }

          return renderLayoutComponent(id);
        })}
      </div>
    </Fieldset>
  );
}
