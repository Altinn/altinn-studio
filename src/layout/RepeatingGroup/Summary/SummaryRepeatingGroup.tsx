import React from 'react';

import { ErrorPaper } from 'src/components/message/ErrorPaper';
import { Lang } from 'src/features/language/Lang';
import { useLanguage } from 'src/features/language/useLanguage';
import { useDeepValidationsForNode } from 'src/features/validation/selectors/deepValidationsForNode';
import { hasValidationErrors } from 'src/features/validation/utils';
import { CompCategory } from 'src/layout/common';
import { LargeGroupSummaryContainer } from 'src/layout/RepeatingGroup/Summary/LargeGroupSummaryContainer';
import classes from 'src/layout/RepeatingGroup/Summary/SummaryRepeatingGroup.module.css';
import { EditButton } from 'src/layout/Summary/EditButton';
import { SummaryComponent } from 'src/layout/Summary/SummaryComponent';
import { Hidden } from 'src/utils/layout/NodesContext';
import { useNodeDirectChildren, useNodeItem } from 'src/utils/layout/useNodeItem';
import { typedBoolean } from 'src/utils/typing';
import type { SummaryRendererProps } from 'src/layout/LayoutComponent';
import type { RepGroupRow, RepGroupRows } from 'src/layout/RepeatingGroup/types';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';

interface FullProps extends SummaryRendererProps<'RepeatingGroup'> {
  rows: RepGroupRows;
  inExcludedChildren: (n: LayoutNode) => boolean;
}

interface FullRowProps extends Omit<FullProps, 'rows'> {
  row: RepGroupRow;
}

export function SummaryRepeatingGroup(props: SummaryRendererProps<'RepeatingGroup'>) {
  const { excludedChildren, largeGroup } = useNodeItem(props.summaryNode) ?? {};
  const { rows: _rows } = useNodeItem(props.targetNode);

  const inExcludedChildren = (n: LayoutNode) =>
    excludedChildren ? excludedChildren.includes(n.id) || excludedChildren.includes(n.baseId) : false;

  const rows: RepGroupRow[] = [];
  for (const row of _rows) {
    if (!row || row.groupExpressions?.hiddenRow || row.index === undefined) {
      continue;
    }
    rows.push(row);
  }

  if (largeGroup && props.overrides?.largeGroup !== false && rows.length) {
    return (
      <LargeRepeatingGroup
        {...props}
        rows={rows}
        inExcludedChildren={inExcludedChildren}
      />
    );
  }

  return (
    <RegularRepeatingGroup
      {...props}
      rows={rows}
      inExcludedChildren={inExcludedChildren}
    />
  );
}

function RegularRepeatingGroup(props: FullProps) {
  const { onChangeClick, changeText, summaryNode, targetNode, overrides, rows } = props;

  const { display: summaryDisplay } = useNodeItem(summaryNode) ?? {};
  const { textResourceBindings: trb } = useNodeItem(targetNode);

  const display = overrides?.display || summaryDisplay;
  const { langAsString } = useLanguage(targetNode);

  const groupValidations = useDeepValidationsForNode(targetNode);
  const groupHasErrors = hasValidationErrors(groupValidations);

  const summaryAccessibleTitleTrb = trb && 'summaryAccessibleTitle' in trb ? trb.summaryAccessibleTitle : undefined;
  const summaryTitleTrb = trb && 'summaryTitle' in trb ? trb.summaryTitle : undefined;
  const titleTrb = trb && 'title' in trb ? trb.title : undefined;
  const ariaLabel = langAsString(summaryTitleTrb ?? summaryAccessibleTitleTrb ?? titleTrb);

  return (
    <>
      <div
        data-testid={'summary-group-component'}
        style={{ width: '100%' }}
      >
        <div className={classes.container}>
          <span className={classes.label}>
            <Lang
              id={summaryTitleTrb ?? titleTrb}
              node={targetNode}
            />
          </span>
          {!display?.hideChangeButton ? (
            <EditButton
              onClick={onChangeClick}
              editText={changeText}
              label={ariaLabel}
            />
          ) : null}
        </div>
        <div style={{ width: '100%' }}>
          {rows.length === 0 ? (
            <span className={classes.emptyField}>
              <Lang id={'general.empty_summary'} />
            </span>
          ) : (
            rows.filter(typedBoolean).map((row) => (
              <RegularRepeatingGroupRow
                key={`row-${row.uuid}`}
                {...props}
                row={row}
              />
            ))
          )}
        </div>
      </div>

      {groupHasErrors && !display?.hideValidationMessages && (
        <div className={classes.gridStyle}>
          <ErrorPaper message={<Lang id={'group.row_error'} />} />
          <div>
            {!display?.hideChangeButton && (
              <button
                className={classes.link}
                onClick={onChangeClick}
                type='button'
              >
                <Lang id={'form_filler.summary_go_to_correct_page'} />
              </button>
            )}
          </div>
        </div>
      )}
    </>
  );
}

function RegularRepeatingGroupRow({
  targetNode,
  inExcludedChildren,
  row,
  onChangeClick,
  changeText,
  summaryNode,
}: FullRowProps) {
  const isHidden = Hidden.useIsHiddenSelector();
  const children = useNodeDirectChildren(targetNode, row.index);

  const childSummaryComponents = children
    .filter((n) => !inExcludedChildren(n))
    .map((child) => {
      if (!isHidden(child) && child.isCategory(CompCategory.Form)) {
        return { component: child.def.renderCompactSummary.bind(child.def), child };
      }
    })
    .filter(typedBoolean);

  return (
    <div
      data-testid={'summary-repeating-row'}
      key={`row-${row.uuid}`}
      className={classes.border}
    >
      {childSummaryComponents.map(({ component: RenderCompactSummary, child }) => (
        <RenderCompactSummary
          onChangeClick={onChangeClick}
          changeText={changeText}
          key={child.id}
          targetNode={child as never} // FIXME: Never type
          summaryNode={summaryNode}
          overrides={{}}
        />
      ))}
    </div>
  );
}

function LargeRepeatingGroup({ targetNode, summaryNode, overrides, inExcludedChildren, rows }: FullProps) {
  const isHidden = Hidden.useIsHiddenSelector();

  return (
    <>
      {rows.filter(typedBoolean).map((row) => (
        <LargeGroupSummaryContainer
          key={`summary-${targetNode.id}-${row.uuid}`}
          id={`summary-${targetNode.id}-${row.index}`}
          groupNode={targetNode}
          restriction={row.index}
          renderLayoutNode={(n) => {
            if (inExcludedChildren(n) || isHidden(n)) {
              return null;
            }

            return (
              <SummaryComponent
                key={n.id}
                summaryNode={summaryNode}
                overrides={{
                  ...overrides,
                  targetNode: n,
                  grid: {},
                  largeGroup: false,
                }}
              />
            );
          }}
        />
      ))}
    </>
  );
}
