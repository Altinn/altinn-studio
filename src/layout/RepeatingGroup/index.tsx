import React, { forwardRef } from 'react';
import type { JSX } from 'react';

import type { PropsFromGenericComponent, ValidateComponent, ValidationFilter, ValidationFilterFunction } from '..';

import { FrontendValidationSource, ValidationMask } from 'src/features/validation';
import { RepeatingGroupDef } from 'src/layout/RepeatingGroup/config.def.generated';
import { GroupHierarchyGenerator } from 'src/layout/RepeatingGroup/hierarchy';
import { RepeatingGroupContainer } from 'src/layout/RepeatingGroup/RepeatingGroupContainer';
import { RepeatingGroupProvider } from 'src/layout/RepeatingGroup/RepeatingGroupContext';
import { RepeatingGroupsFocusProvider } from 'src/layout/RepeatingGroup/RepeatingGroupFocusContext';
import { SummaryRepeatingGroup } from 'src/layout/RepeatingGroup/Summary/SummaryRepeatingGroup';
import type { LayoutValidationCtx } from 'src/features/devtools/layoutValidation/types';
import type { BaseValidation, ComponentValidation } from 'src/features/validation';
import type { SummaryRendererProps } from 'src/layout/LayoutComponent';
import type { ComponentHierarchyGenerator } from 'src/utils/layout/HierarchyGenerator';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';

export class RepeatingGroup extends RepeatingGroupDef implements ValidateComponent, ValidationFilter {
  private _hierarchyGenerator = new GroupHierarchyGenerator();

  directRender(): boolean {
    return true;
  }

  render = forwardRef<HTMLDivElement, PropsFromGenericComponent<'RepeatingGroup'>>(
    function LayoutComponentRepeatingGroupRender(props, ref): JSX.Element | null {
      return (
        <RepeatingGroupProvider node={props.node}>
          <RepeatingGroupsFocusProvider>
            <RepeatingGroupContainer ref={ref} />
          </RepeatingGroupsFocusProvider>
        </RepeatingGroupProvider>
      );
    },
  );

  renderSummary({
    onChangeClick,
    changeText,
    summaryNode,
    targetNode,
    overrides,
  }: SummaryRendererProps<'RepeatingGroup'>): JSX.Element | null {
    return (
      <SummaryRepeatingGroup
        onChangeClick={onChangeClick}
        changeText={changeText}
        summaryNode={summaryNode}
        targetNode={targetNode}
        overrides={overrides}
      />
    );
  }

  renderSummaryBoilerplate(): boolean {
    return false;
  }

  getDisplayData(): string {
    return '';
  }

  hierarchyGenerator(): ComponentHierarchyGenerator<'RepeatingGroup'> {
    return this._hierarchyGenerator;
  }

  runComponentValidation(node: LayoutNode<'RepeatingGroup'>): ComponentValidation[] {
    if (!node.item.dataModelBindings) {
      return [];
    }

    const validations: ComponentValidation[] = [];
    // check if minCount is less than visible rows
    const repeatingGroupComponent = node.item;
    const repeatingGroupMinCount = repeatingGroupComponent.minCount ?? 0;
    const repeatingGroupVisibleRows = repeatingGroupComponent.rows.filter(
      (row) => row && !row.groupExpressions?.hiddenRow,
    ).length;

    // Validate minCount
    if (repeatingGroupVisibleRows < repeatingGroupMinCount) {
      validations.push({
        message: { key: 'validation_errors.minItems', params: [repeatingGroupMinCount] },
        severity: 'error',
        componentId: node.item.id,
        source: FrontendValidationSource.Component,
        // Treat visibility of minCount the same as required to prevent showing an error immediately
        category: ValidationMask.Required,
      });
    }

    return validations;
  }

  /**
   * Repeating group has its own minCount property, so if set, we should filter out the minItems validation from schema.
   */
  private schemaMinItemsFilter(validation: BaseValidation): boolean {
    return !(
      validation.source === FrontendValidationSource.Schema && validation.message.key === 'validation_errors.minItems'
    );
  }

  getValidationFilters(_node: LayoutNode<'RepeatingGroup'>): ValidationFilterFunction[] {
    if ((_node.item.minCount ?? 0) > 0) {
      return [this.schemaMinItemsFilter];
    }
    return [];
  }

  isDataModelBindingsRequired(): boolean {
    return true;
  }

  validateDataModelBindings(ctx: LayoutValidationCtx<'RepeatingGroup'>): string[] {
    const [errors, result] = this.validateDataModelBindingsAny(ctx, 'group', ['array']);
    if (errors) {
      return errors;
    }

    if (result) {
      const innerType = Array.isArray(result.items) ? result.items[0] : result.items;
      if (!innerType || typeof innerType !== 'object' || !innerType.type || innerType.type !== 'object') {
        return [`group-datamodellbindingen peker mot en ukjent type i datamodellen`];
      }
    }

    return [];
  }
}
