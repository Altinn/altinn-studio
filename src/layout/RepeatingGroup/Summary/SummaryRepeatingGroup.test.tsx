import React from 'react';

import { jest } from '@jest/globals';

import { ALTINN_ROW_ID } from 'src/features/formData/types';
import { SummaryRepeatingGroup } from 'src/layout/RepeatingGroup/Summary/SummaryRepeatingGroup';
import { renderWithNode } from 'src/test/renderWithProviders';
import { useNode } from 'src/utils/layout/NodesContext';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';

describe('SummaryGroupComponent', () => {
  let mockHandleDataChange: () => void;

  beforeEach(() => {
    mockHandleDataChange = jest.fn();
  });

  test('SummaryGroupComponent -- should match snapshot', async () => {
    const { asFragment } = await render();
    expect(asFragment()).toMatchSnapshot();
  });

  function TestComponent({ node, groupId }: { node: LayoutNode<'Summary'>; groupId: string }) {
    const groupNode = useNode(groupId) as LayoutNode<'RepeatingGroup'>;
    return (
      <SummaryRepeatingGroup
        changeText={'Change'}
        onChangeClick={mockHandleDataChange}
        summaryNode={node}
        targetNode={groupNode}
      />
    );
  }

  async function render() {
    return await renderWithNode<true, LayoutNode<'Summary'>>({
      nodeId: 'mySummary',
      inInstance: true,
      renderer: ({ node }) => (
        <TestComponent
          node={node}
          groupId='groupComponent'
        />
      ),
      queries: {
        fetchFormData: async () => ({
          mockGroup: [
            {
              [ALTINN_ROW_ID]: 'abc123',
              mockDataBinding1: '1',
              mockDataBinding2: '2',
            },
          ],
        }),
        fetchLayouts: async () => ({
          FormLayout: {
            data: {
              layout: [
                {
                  type: 'RepeatingGroup',
                  id: 'groupComponent',
                  dataModelBindings: {
                    group: 'mockGroup',
                  },
                  textResourceBindings: {
                    title: 'mockGroupTitle',
                  },
                  children: ['0:mockId1', '1:mockId2'],
                  edit: {
                    multiPage: true,
                  },
                  maxCount: 3,
                },
                {
                  type: 'Input',
                  id: 'mockId1',
                  dataModelBindings: {
                    simpleBinding: 'mockGroup.mockDataBinding1',
                  },
                  readOnly: false,
                  required: false,
                  textResourceBindings: {
                    title: 'mockField1',
                  },
                  triggers: [],
                },
                {
                  type: 'Input',
                  id: 'mockId2',
                  dataModelBindings: {
                    simpleBinding: 'mockGroup.mockDataBinding2',
                  },
                  readOnly: false,
                  required: false,
                  textResourceBindings: {
                    title: 'mockField2',
                  },
                  triggers: [],
                },
                {
                  type: 'Summary',
                  id: 'mySummary',
                  componentRef: 'groupComponent',
                  largeGroup: false,
                },
              ],
            },
          },
        }),
        fetchTextResources: () =>
          Promise.resolve({
            language: 'nb',
            resources: [
              { id: 'mockGroupTitle', value: 'Mock group' },
              { id: 'mockField1', value: 'Mock field 1' },
              { id: 'mockField2', value: 'Mock field 2' },
            ],
          }),
      },
    });
  }
});
