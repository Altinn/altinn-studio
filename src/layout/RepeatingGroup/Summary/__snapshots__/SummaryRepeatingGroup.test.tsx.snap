// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryRepeatingGroup SummaryRepeatingGroup -- should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: none;"
  >
     
  </div>
  <div
    data-testid="summary-group-component"
    style="width: 100%;"
  >
    <div
      class="container"
    >
      <span
        class="label"
      >
        <span>
          Mock group
        </span>
      </span>
      <button
        aria-label="Change: Mock group"
        class="fds-btn fds-focus fds-btn--sm fds-btn--tertiary fds-btn--second editButton"
        type="button"
      >
        Change
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          font-size="1rem"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M19.638 4.417a3.25 3.25 0 0 0-4.608-.008l-9.378 9.379a.75.75 0 0 0-.191.324l-1.414 4.95a.75.75 0 0 0 .925.927l4.94-1.398a.75.75 0 0 0 .327-.191l9.39-9.391a3.25 3.25 0 0 0 .01-4.592M16.091 5.47a1.752 1.752 0 1 1 2.478 2.478l-.23.23-2.477-2.479zM14.8 6.76 6.85 14.71l-.991 3.47 3.457-.979 7.963-7.963z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
    <div
      style="width: 100%;"
    >
      <div
        class="border"
        data-testid="summary-repeating-row"
      >
        <div
          data-testid="summary-item-compact"
        >
          <span>
            Mock field 1 : 
          </span>
          <span
            class="data"
          >
            1
          </span>
        </div>
        <div
          data-testid="summary-item-compact"
        >
          <span>
            Mock field 2 : 
          </span>
          <span
            class="data"
          >
            2
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
